Armadillo/src/main/java/com/scribd/armadillo/playback/mediasource/DashMediaSourceGenerator.kt

```diff
@@ -1,36 +1,63 @@
 package com.scribd.armadillo.playback.mediasource
 
 import android.content.Context
+import android.os.Handler
 import com.google.android.exoplayer2.drm.DrmSessionManagerProvider
 import com.google.android.exoplayer2.offline.Download
 import com.google.android.exoplayer2.offline.DownloadHelper
 import com.google.android.exoplayer2.source.MediaSource
 import com.google.android.exoplayer2.source.dash.DashMediaSource
+import com.scribd.armadillo.StateStore
+import com.scribd.armadillo.actions.OpeningLicenseAction
 import com.scribd.armadillo.download.DownloadTracker
+import com.scribd.armadillo.download.drm.events.WidevineSessionEventListener
 import com.scribd.armadillo.extensions.toUri
 import com.scribd.armadillo.models.AudioPlayable
+import com.scribd.armadillo.models.DrmType
 import javax.inject.Inject
 
+/** For playback, both streaming and downloaded */
 internal class DashMediaSourceGenerator @Inject constructor(
+    context: Context,
     private val mediaSourceHelper: HeadersMediaSourceHelper,
     private val downloadTracker: DownloadTracker,
     private val drmMediaSourceHelper: DrmMediaSourceHelper,
     private val drmSessionManagerProvider: DrmSessionManagerProvider,
+    private val stateStore: StateStore.Modifier,
 ) : MediaSourceGenerator {
 
+    private val drmHandler = Handler(context.mainLooper)
+
     override fun generateMediaSource(context: Context, request: AudioPlayable.MediaRequest): MediaSource {
+        if (request.drmInfo != null) {
+            stateStore.dispatch(OpeningLicenseAction(request.drmInfo.drmType))
+        }
         val dataSourceFactory = mediaSourceHelper.createDataSourceFactory(context, request)
 
-        downloadTracker.getDownload(request.url.toUri())?.let {
-            if (it.state != Download.STATE_FAILED) {
-                val mediaItem = drmMediaSourceHelper.createMediaItem(context = context, request = request, isDownload = true)
-                return DownloadHelper.createMediaSource(it.request, dataSourceFactory, drmSessionManagerProvider.get(mediaItem))
-            }
-        }
+        val download = downloadTracker.getDownload(request.url.toUri())
+        val isDownloaded = download != null && download.state != Download.STATE_FAILED
+        val mediaItem = drmMediaSourceHelper.createMediaItem(context = context, request = request, isDownload = isDownloaded)
 
-        val mediaItem = drmMediaSourceHelper.createMediaItem(context = context, request = request, isDownload = false)
-        return DashMediaSource.Factory(dataSourceFactory)
-            .createMediaSource(mediaItem)
+        return if (isDownloaded) {
+            val drmManager = drmSessionManagerProvider.get(mediaItem)
+            DownloadHelper.createMediaSource(download!!.request, dataSourceFactory, drmManager)
+        } else {
+            DashMediaSource.Factory(dataSourceFactory)
+                .setDrmSessionManagerProvider(drmSessionManagerProvider)
```

Comment:
Should this also only get added if there's DRM, or does it not matter?

Reply:
should not matter